### PR TITLE
Add validation to http_prefix setting

### LINKED ISSUE #1927: -http.prefix with empty string causes 404
With a Prometheus server running on localhost:
```
$ curl localhost:9090
<a href="/graph">Found</a>.
```

And setting up the cortex query-frontend to use it:
```
$ ./cortex \
   -target=query-frontend \
   -frontend.downstream-url=http://localhost:9090 \
   -server.http-listen-port=9091 \
   -auth.enabled=false \
   -http.prefix=
level=info ts=2019-12-19T14:48:56.894668Z caller=cortex.go:232 msg=initialising module=server
level=info ts=2019-12-19T14:48:56.894919Z caller=server.go:117 http=[::]:9091 grpc=[::]:9095 msg="server listening on addresses"
level=info ts=2019-12-19T14:48:56.895159Z caller=cortex.go:232 msg=initialising module=overrides
level=info ts=2019-12-19T14:48:56.895184Z caller=override.go:53 msg="per-tenant overrides disabled"
level=info ts=2019-12-19T14:48:56.895201Z caller=cortex.go:232 msg=initialising module=query-frontend
level=info ts=2019-12-19T14:48:56.89525Z caller=main.go:76 msg="Starting Cortex" version="(version=, branch=, revision=)"
```

everything is cushty:
```
$ curl localhost:9091
<a href="/graph">Found</a>.
```

However, if the empty string is used for the `-http.prefix`, as in:
```
$ ./cortex \
   -target=query-frontend \
   -frontend.downstream-url=http://localhost:9090 \
   -server.http-listen-port=9091 \
   -auth.enabled=false \
    -http.prefix='""'
level=info ts=2019-12-19T14:49:40.485835Z caller=cortex.go:232 msg=initialising module=server
level=info ts=2019-12-19T14:49:40.48618Z caller=server.go:117 http=[::]:9091 grpc=[::]:9095 msg="server listening on addresses"
level=info ts=2019-12-19T14:49:40.486473Z caller=cortex.go:232 msg=initialising module=overrides
level=info ts=2019-12-19T14:49:40.486507Z caller=override.go:53 msg="per-tenant overrides disabled"
level=info ts=2019-12-19T14:49:40.486531Z caller=cortex.go:232 msg=initialising module=query-frontend
level=info ts=2019-12-19T14:49:40.486706Z caller=main.go:76 msg="Starting Cortex" version="(version=, branch=, revision=)"
```

then you only get 404s and everything is sad:
```
$ curl localhost:9091
404 page not found
```

This isn't really an issue on the command line (who the fuck writes `'""`), but in a Kuberentes deployment, it's pretty easy to do it by accident (i know, because I just did it :D), e.g:
```
      - args:
        - -http.prefix=""
```
and then it is pretty much impossible to debug this.

Ideally, `-http.prefix=` and `-http.prefix=""` should be the same, or at least post an error message that `-http.prefix=""` is gonna make you sad

### PR DESCRIPTION
**Pull request by ethervoid**
http_prefix setting doesn't have validation so in case a user sets that
setting to a value without stating with / or empty value that would prevent
cortex to start with a panic coming from the router

Signed-off-by: Mario de Frutos <mario@defrutos.org>

**What this PR does**:

This PR validates the `http_prefix` setting to avoid issues passing not valid values. The original issue talked about not able to reach the enpoint if the settings was `-http.prefix='""'` but right now that returns a panic from the router:

```
panic: path must begin with '/' in path '""/api/v1/*path'

goroutine 1 [running]:
github.com/julienschmidt/httprouter.(*Router).Handle(0xc000a2e720, 0x2603f7b, 0x7, 0xc0001aea80, 0xf, 0xc0000a9c80)
	/home/ethervoid/development/cortex/vendor/github.com/julienschmidt/httprouter/router.go:246 +0x24c
github.com/julienschmidt/httprouter.(*Router).OPTIONS(...)
	/home/ethervoid/development/cortex/vendor/github.com/julienschmidt/httprouter/router.go:213
github.com/prometheus/common/route.(*Router).Options(0xc000a83580, 0x2601b78, 0x6, 0xc000a835c0)
	/home/ethervoid/development/cortex/vendor/github.com/prometheus/common/route/route.go:94 +0xd8
github.com/prometheus/prometheus/web/api/v1.(*API).Register(0xc000252300, 0xc000a83580)
	/home/ethervoid/development/cortex/vendor/github.com/prometheus/prometheus/web/api/v1/api.go:285 +0xb1
github.com/cortexproject/cortex/pkg/api.NewQuerierHandler(0x0, 0x260f53c, 0xd, 0x260b3b0, 0xb, 0x0, 0x0, 0x7ffd6e426706, 0x2, 0x2a54040, ...)
	/home/ethervoid/development/cortex/pkg/api/handlers.go:235 +0xe4b
github.com/cortexproject/cortex/pkg/cortex.(*Cortex).initQuerier(0xc0008e8000, 0x0, 0xc00053b498, 0x0, 0xc00053b440)
	/home/ethervoid/development/cortex/pkg/cortex/modules.go:271 +0xf9
github.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule(0xc000116d60, 0x7ffd6e426690, 0x7, 0xc000ea8ab0, 0xc0008c4c00, 0x0, 0x0)
	/home/ethervoid/development/cortex/pkg/util/modules/modules.go:103 +0x2c2
github.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices(0xc000116d60, 0xc0002c15c0, 0x1, 0x1, 0xc00053b4f0, 0x7510c18f88e2e501, 0xc000782200)
	/home/ethervoid/development/cortex/pkg/util/modules/modules.go:75 +0x105
github.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run(0xc0008e8000, 0xc0006ddb80, 0x4)
	/home/ethervoid/development/cortex/pkg/cortex/cortex.go:369 +0x2a5
main.main()
	/home/ethervoid/development/cortex/cmd/cortex/main.go:190 +0xe70
```

So it's a small change and improves the user experience and prevents possible side effects

**Which issue(s) this PR fixes**:
Fixes #1927

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by ethervoid** [comment]
@pracucci I've checked this issue, and can't reproduce the original issue but if you put the prefix with an invalid value you get a panic (check the PR description) at startup so I think it's ok to include it in the validation. What do you think?

File: pkg/cortex/cortex_test.go [hunk 1]
```diff
@@ -94,3 +94,46 @@ func TestCortex(t *testing.T) {
...
+			name: "should pass validation if the http prefix starts with /",
+			getTestConfig: func() *Config {
+				configuration := newDefaultConfig()
+				configuration.HTTPPrefix = "/test"
+				return configuration
+			},
+			expectedError: false,
+		},
+		{
+			name: "should fail if validation is not empty or starts with /",
```
Comment (pstibrany) [inline pkg/cortex/cortex_test.go:121]:
```suggestion
			name: "should fail validation for invalid prefix",
```

**Review by pstibrany — APPROVED** [review]
Thank you!

File: pkg/cortex/cortex.go [hunk 2]
```diff
@@ -173,6 +174,10 @@ func (c *Config) Validate(log log.Logger) error {
 		return err
 	}
 
+	if c.HTTPPrefix != "" && !strings.HasPrefix(c.HTTPPrefix, "/") {
+		return errors.New("http_prefix should be empty or start with /")
```
Comment (pracucci) [inline pkg/cortex/cortex.go:178]: Define this error in a global variable `errInvalidHttpPrefix` and then use it to check the exact error in the unit test.

File: pkg/cortex/cortex_test.go [hunk 3]
```diff
@@ -94,3 +94,46 @@ func TestCortex(t *testing.T) {
 	// check that compactor is configured which is not part of Target=All
 	require.NotNil(t, serviceMap[Compactor])
 }
+
+func TestConfigValidation(t *testing.T) {
+	for _, tc := range []struct {
+		name          string
+		getTestConfig func() *Config
+		expectedError bool
```
Comment (pracucci) [inline pkg/cortex/cortex_test.go:102]: To have an higher confidence on test correctness, you could check the exact error. We expect `errInvalidHttpPrefix` (see other comment).

```suggestion
		expectedError error
```

File: pkg/cortex/cortex.go (hunk 2 above)
Comment (pracucci) [inline pkg/cortex/cortex.go:178]: I wouldn't mention the YAML config name in the error, considering we also support CLI flags. I would just be generic (it's the pattern we've used in most validation errors).

```suggestion
		return errors.New("HTTP prefix should be empty or start with /")
```

**Review by pracucci — APPROVED** [review]
LGTM, thanks! I left few nits, but consider it already approved.

File: pkg/cortex/cortex.go (hunk 2 above)
Comment (ethervoid) [reply pkg/cortex/cortex.go]: Done! Yeah makes a lot of sense :)

**Comment by ethervoid** [comment]
Changes made, ready to merge thank you for the CR and the feedback :-)

**Pull request merged**